microsetta_admin/server.py

```diff
@@ -512,35 +514,117 @@ def new_kits():
+                barcode_file = barcode_file.read().decode('utf-8')
+                new_barcodes = [line.split(',')[0].strip('\ufeff\r')
+                                for line in barcode_file.split('\n')
+                                if line.strip()]
+                if len(new_barcodes) != num_kits:
+                    if len(new_barcodes) > num_kits:
+                        error_message = (f'Too many barcodes provided in '
+                                         f'file {barcode_file}.')
+                    else:
+                        error_message = (f'Not enough barcodes provided in '
+                                         f'file {barcode_file}.')
+                    return render_template('create_kits.html',
+                                           error_message=error_message,
+                                           projects=projects,
+                                           **build_login_variables())
+
+                user_barcodes.append(new_barcodes)
+                total_provided_samples += len(new_barcodes)
+
+        # Calculate remaining samples to be generated
+        total_needed_samples = num_kits * num_samples
+        remaining_samples_to_generate = total_needed_samples - \
+            total_provided_samples
+
+        if remaining_samples_to_generate < 0:
+            error_message = (f'Too many samples provided '
+                             f'({total_provided_samples}) '
+                             f'for the specified number of kits and samples '
+                             f'({total_needed_samples}).')
+            return render_template('create_kits.html',
+                                   error_message=error_message,
+                                   projects=projects,
+                                   **build_login_variables())
+
+        payload = {
+            'action': 'create' if not user_barcodes else 'insert',
+            'project_ids': selected_project_ids,
+            'number_of_kits': num_kits,
+            'number_of_samples': num_samples,
+            'generate_barcodes': generate_barcodes,
+            'user_barcodes': user_barcodes
+        }
+
         if prefix:
             payload['kit_id_prefix'] = prefix
+        if user_barcodes:
+            total_user_barcodes = sum(len(slot) for slot in user_barcodes)
+
+            if total_needed_samples != sum(len(slot) for slot in
+                                           user_barcodes) \
+                    + remaining_samples_to_generate:
+                error_message = (f'The total number of kit IDs '
+                                 f'({total_needed_samples}) '
+                                 f'does not match the sum of '
+                                 f'provided barcodes '
+                                 f'({total_user_barcodes})'
+                                 f' and remaining samples to generate '
+                                 f'({remaining_samples_to_generate}).')
```

Comment:
Is it still possible to hit the block of code since you're validating the length of each provided CSV now?

Reply:
You're right, it's not possible since it's being checked with `if len(new_barcodes) != num_kits`. Thank you, removing.